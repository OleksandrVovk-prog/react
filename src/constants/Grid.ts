export const GRID_WIDTH = 50;
export const GRID_HEIGHT = 50;
