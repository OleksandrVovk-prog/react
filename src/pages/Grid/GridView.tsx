import { type ReactElement, useEffect } from 'react';
import Draggable from 'react-draggable';
import { toast, ToastContainer } from 'react-toastify';

import { GRID_HEIGHT, GRID_WIDTH } from '../../constants/Grid';
import { getContrastColor } from '../../utils/helpers';

import type { IGridView } from './interfaces/IGridView';

import styles from './sass/GridView.module.scss';

function GridView({
  squares,
  contextMenu,
  onHandleRightClick,
  onHandleCreateSquare,
  onHideContextMenu,
  errorMessage,
  setErrorMessage,
  handleDragStop,
  handleDrag,
}: IGridView): ReactElement {
  useEffect(() => {
    if (errorMessage) {
      toast.error(errorMessage, {
        onClose: () => {
          setErrorMessage('');
        },
      });
    }
  }, [errorMessage]);

  return (
    <div
      aria-hidden
      className={styles.gridContainer}
      onClick={onHideContextMenu}
      onContextMenu={onHandleRightClick}
      style={{ backgroundSize: `${GRID_WIDTH}px ${GRID_HEIGHT}px` }}
    >
      <ToastContainer
        position="top-right"
        autoClose={1000}
        hideProgressBar
        newestOnTop={false}
        closeOnClick
        rtl={false}
        pauseOnFocusLoss
        draggable
        pauseOnHover
      />
      {contextMenu ? (
        <div
          className={styles.gridContextMenu}
          style={{
            top: contextMenu.y,
            left: contextMenu.x,
          }}
        >
          <div aria-hidden onClick={onHandleCreateSquare}>
            Create Square
          </div>
        </div>
      ) : null}
      {squares?.map((square) => (
        <Draggable
          key={square.id}
          defaultPosition={{ x: square.x, y: square.y }}
          position={{ x: square.x, y: square.y }}
          grid={[GRID_WIDTH, GRID_HEIGHT]}
          onStop={(e, data) => handleDragStop(square.id, e, data)}
          onDrag={(e, data) => handleDrag(square.id, e, data)}
        >
          <div
            className={styles.gridItem}
            style={{
              width: GRID_WIDTH,
              height: GRID_HEIGHT,
              backgroundColor: square.color,
              color: getContrastColor(square.color),
            }}
          >
            Eddie
          </div>
        </Draggable>
      ))}
    </div>
  );
}

export default GridView;
