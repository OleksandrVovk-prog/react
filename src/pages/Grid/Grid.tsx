import React, { useState } from 'react';

import { GRID_HEIGHT, GRID_WIDTH } from '../../constants/Grid';
import { getItemSize, getRandomColor } from '../../utils/helpers';
import GridView from './GridView';

import type { ReactElement } from 'react';
import type { DraggableData, DraggableEvent } from 'react-draggable';
import type { IContextMenu } from './interfaces/IContextMenu';
import type { ISquare } from './interfaces/ISquare';
import type { ISquareLastPosition } from './interfaces/ISquareLastPosition';

function Grid(): ReactElement {
  const [squares, setSquares] = useState<ISquare[]>([]);
  const [contextMenu, setContextMenu] = useState<IContextMenu | null>(null);
  const [errorMessage, setErrorMessage] = useState<string>('');
  const [lastPositionMoved, setLastPositionMoved] = useState<ISquareLastPosition | null>(null);

  const onHideContextMenu = () => {
    setContextMenu(null);
  };

  const onHandleRightClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setContextMenu({ x: e.clientX, y: e.clientY });
  };

  // eslint-disable-next-line arrow-body-style
  const checkCollision = (id: number, newX: number, newY: number): boolean => {
    return squares.some((square) => square.id !== id && square.x === newX && square.y === newY);
  };

  const onHandleCreateSquare = () => {
    if (contextMenu) {
      const xAxis = getItemSize(contextMenu.x, GRID_WIDTH);
      const yAxis = getItemSize(contextMenu.y, GRID_HEIGHT);
      const isAlreadyAdded = squares.some((square) => square.x === xAxis && square.y === yAxis);
      if (!isAlreadyAdded) {
        setSquares([...squares, {
          id: Date.now(), x: xAxis, y: yAxis, color: getRandomColor(),
        }]);
        setContextMenu(null);
      } else {
        setErrorMessage('Square already exists');
        setContextMenu(null);
      }
    }
  };

  const handleDragStop = (id: number, e: DraggableEvent, data: DraggableData) => {
    const newX = getItemSize(data.x, GRID_WIDTH);
    const newY = getItemSize(data.y, GRID_HEIGHT);
    if (checkCollision(id, newX, newY)) {
      if (id === lastPositionMoved?.id) {
        setSquares((prev) => prev.map(
          (sq) => (sq.id === id ? { ...sq, x: lastPositionMoved.lastX, y: lastPositionMoved.lastY } : sq),
        ));
      }
      setErrorMessage('Square in this position already exists');
    } else {
      setSquares((prev) => prev.map((sq) => (sq.id === id ? { ...sq, x: newX, y: newY } : sq)));
    }
    setLastPositionMoved(null);
  };

  const handleDrag = (id: number, e: DraggableEvent, data: DraggableData) => {
    const { lastX, lastY } = data;
    setLastPositionMoved({ id, lastX, lastY });
  };

  return (
    <GridView
      squares={squares}
      contextMenu={contextMenu}
      onHandleRightClick={onHandleRightClick}
      onHandleCreateSquare={onHandleCreateSquare}
      onHideContextMenu={onHideContextMenu}
      errorMessage={errorMessage}
      setErrorMessage={setErrorMessage}
      handleDragStop={handleDragStop}
      handleDrag={handleDrag}
    />
  );
}

export default Grid;
