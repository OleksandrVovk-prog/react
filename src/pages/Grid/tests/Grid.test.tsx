import { render } from '../../../../__tests__/test-utils';
import Grid from '../Grid';

describe('<Grid />', () => {
  test('should render component', async () => {
    const { asFragment } = render(<Grid />);
    expect(asFragment()).toMatchSnapshot();
  });
});
