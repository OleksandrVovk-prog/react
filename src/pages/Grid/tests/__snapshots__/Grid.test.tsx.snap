// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> should render component 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="gridContainer"
    style="background-size: 50px 50px;"
  >
    <section
      aria-atomic="false"
      aria-label="Notifications Alt+T"
      aria-live="polite"
      aria-relevant="additions text"
      class="Toastify"
    />
  </div>
</DocumentFragment>
`;
